File: PA/LabsPA/Lab1/java/task-3/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;
import java.lang.Math;;


public class Main {
	static class Task {
		public final static String INPUT_FILE = "in";
		public final static String OUTPUT_FILE = "out";

		int n, x, y;

		private void readInput() {
			try {
				Scanner sc = new Scanner(new File(INPUT_FILE));
				n = sc.nextInt();
				x = sc.nextInt();
				y = sc.nextInt();
				sc.close();
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}

		private void writeOutput(int answer) {
			try {
				PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
				pw.printf("%d\n", answer);
				pw.close();
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}

		int N = (int)Math.pow(2, n);
		int count;
		int[][] grid = new int[N + 1][N + 1];

		private void parcurgere(int i, int  j, int n) {
			if (n > 0) {
				int offset = (int)Math.pow(2, n - 1);
				parcurgere(i, j, n - 1);
				parcurgere(i, j + offset, n - 1);
				parcurgere(i + offset, j, n - 1);
				parcurgere(i + offset, j + offset, n - 1);
			} else {
				grid[i][j] = count;
				count++;
			}
		}

		private int getAnswer(int n, int x, int y) {
			// TODO: Calculati valoarea de pe pozitia (x, y) din matricea de dimensiune
			// 2^N * 2^N.
			count = 1;
			parcurgere(1, 1, n);

			return grid[x][y];
		}

		public void solve() {
			readInput();
			writeOutput(getAnswer(n, x, y));
		}
	}

	public static void main(String[] args) {
		new Task().solve();
	}
}

File: PA/LabsPA/Lab1/java/cpp/task-3/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <math.h>

using namespace std;

class Task {
public:
	void solve() {
		read_input();
		print_output(get_result(n, x, y));
	}

private:
	int n, x, y;

	void read_input() {
		ifstream fin("in");
		fin >> n >> x >> y;
		fin.close();
	}

	int get_result(int n, int x, int y) {
		/*
		TODO: Calculati valoarea de pe pozitia (x, y) din matricea de dimensiune
		2^n x 2^n
		*/
		int x1 = 1, x2 = pow(2, n), y1 = 1, y2 = pow(2, n);
		int k = 1;
		int start = 1, end = pow(2, n+n);
		while(k <= n) {
			int aux = (end - start + 1) / 4;
			if(x >= x1 && x <= (x1 + x2)/2 && y >= y1 && y <= (y1 + y2)/2) {
				x2 = (x1 + x2)/2; y2 = (y1 + y2)/2;
				end = aux;
				cout << 1 << '\n';
			} else if (x >= x1 && x <= (x1 + x2)/2 && y >= (y1 + y2)/2 && y <= y2) {
				x2 = (x1 + x2)/2; y1 = (y1 + y2)/2;
				start = aux + 1;
				end = 2 * aux;
				cout << 2 << '\n';
			} else if (x >= (x1 + x2)/2 && x <= x2 && y >= y1 && y <= (y1 + y2) / 2) {
				x1 = (x1 + x2)/2; y2 = (y1 + y2)/2;
				start = 2 * aux + 1;
				end = 3 *aux;
				cout << 3 << '\n';
			} else if (x >= (x1 + x2)/2 && x <= x2 && y >= (y1 + y2)/2 && y <= y2) {
				x1 = (x1 + x2)/2; y1 = (y1 + y2)/2;
				start = 3* aux +1;
				end = 4* aux;
				cout << 4 << '\n';
			}
			k++;
			cout << x1 << " " << x2 << " " << y1 << " " << y2 << '\n'; 
		}
		cout << start <<" " << end;
		
		return start;
	}

	void print_output(int result) {
		ofstream fout("out");
		fout << result;
		fout.close();
	}
};

int main() {
	Task task;
	task.solve();
	return 0;
}
